hydromt/models/region/region.py

```diff
@@ -0,0 +1,275 @@
+        elif "bbox" in region:
+            bbox = region["bbox"]
+            geom = gpd.GeoDataFrame(geometry=[box(*bbox)], crs=4326)
+        elif "geom" in region:
+            geom = region["geom"]
+            if geom.crs is None:
+                raise ValueError('Model region "geom" has no CRS')
+        elif "grid" in region:  # Grid specific - should be removed in the future
+            geom = region["grid"].raster.box
+        elif "model" in region:
+            geom = region["model"].region
+        else:
+            raise ValueError(f"model region argument not understood: {region}")
+
+        self._data = geom
+        # This setup method returns region so that it can be wrapped for models which
+        # require more information, e.g. grid RasterDataArray or xy coordinates.
+        return region
+
+    # def _rasterdataset_to_geom(self, raster_dataset) -> GeoDataFrame:
+    # crs = raster_dataset.raster.crs
+    # coord_index = cast(pd.Index, raster_dataset.raster.coords.to_index())
+    # dims_max = cast(np.ndarray, coord_index.max())
+    # dims_min = cast(np.ndarray, coord_index.min())
+
+    # # in raster datasets it is guaranteed that y_dim is penultimate dim and x_dim is last dim
+    # geom: GeoDataFrame = GeoDataFrame(
+    #     geometry=[
+    #         box(
+    #             xmin=dims_min[-1],
+    #             ymin=dims_min[-2],
+    #             xmax=dims_max[-1],
+    #             ymax=dims_max[-2],
+    #         )
+    #     ],
+    #     crs=crs,
+    # )
+    # return geom
+
+    def set(self, data: GeoDataFrame, kind: str = "user"):
+        """Set the model region based on provided GeoDataFrame."""
+        # if nothing is provided, record that the region was set by the user
+        self._kind = kind
+        self._data = data
+
+    @property
+    def total_bounds(self):
+        """Return the total bound sof the model region."""
+        return self.data.total_bounds
+
+    @property
+    def data(self) -> GeoDataFrame:
+        """Provide access to the underlying data of the model region."""
+        if self._data is None:
+            self.read(DEFAULT_REGION_FILE_PATH)
+
+        return cast(GeoDataFrame, self._data)
+
+    def _get_root(self) -> ModelRoot:
+        # cast is necessary because technically the model could have been
+        # dealocated, but it shouldn't be in this case.
+        root: Optional[ModelRoot] = cast("Model", self.model_ref()).root
+
+        # cannot read geom files for purely in memory models
+        if root is None:
+            raise ValueError("Root was not set, cannot read region file")
+
+        return root
+
+    @property
+    def crs(self) -> CRS:
+        """Provide access to the CRS of the model region."""
+        return self.data.crs
+
+    def read(
+        self,
+        rel_path: StrPath = Path(DEFAULT_REGION_FILE_PATH),
+        **read_kwargs,
+    ):
+        """Read the model region from a file on disk."""
+        if self._data is None:
```

Comment:
In general I would expect if I call the read method that it would always read, not only if it has not yet been initialized. 

See this PR for the implementation the `_initialize_<components>`: https://github.com/Deltares/hydromt/pull/695/files

Also fine to park this for a later PR. It would be good to streamline this behavior between all components.